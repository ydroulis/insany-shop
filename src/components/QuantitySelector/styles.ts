"use client"

import styled from 'styled-components'

export const QuantityContainer = styled.div`
    width: 65px;
    height: 40px;

    color: #737380;
   
    .sr-only {
        position: absolute;
        width: 1px;
        height: 1px;
        padding: 0;
        margin: -1px;
        overflow: hidden;
        clip: rect(0, 0, 0, 0);
        white-space: nowrap;
        border-width: 0;
    }
`

export const QuantitySelector = styled.button`
    display: flex;
    align-items: center;
    justify-content: space-between;

    color: #737380;

    width: 100%;
    height: 100%;
    
    background-color: #F3F5F6;
    border-width: 0px, 0px, 0px, 0px;

    border-style: solid;
    border-color: #A8A8B3;
    border-radius: 8px;

    cursor: pointer;

    padding: 8px 12px 8px 12px;
`

export const QuantitySelectorMenu = styled.div<{ $isOpen: boolean }>`
    display: ${({ $isOpen }) => ($isOpen ? 'flex' : 'none')};
    flex-direction: column;
    gap: 4px;
    margin-top: .25rem;

    width: 100%;

    border-radius: 4px;
    background-color: #FFFFFF;
    position: relative;
    z-index: 99;
`

export const QuantitySelectorItem = styled.button`
    border: none;
    background-color: transparent;
    display: flex;
    align-items: center;
    justify-content: center;
    margin: .5rem 0rem;

    cursor: pointer;

    height: 24px;
    width: 100%;

    margin: 0rem;

    color: #737380;

    font-family: 'Inter', sans-serif;

    text-align: center;
`               