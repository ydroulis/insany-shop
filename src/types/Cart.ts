
export type CartProduct = {
    id: number
    name: string
    description: string
    price: number
    image: string
    stock: number
    items?: number
}

export type Cart = {
    content: CartProduct[]
    total: number
}