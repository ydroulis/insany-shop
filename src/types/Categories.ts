export type Categories = Category[];

export type Category = {
    id: string,
    name: string,
    description: string,
    icon: string,
    productCount: number
};