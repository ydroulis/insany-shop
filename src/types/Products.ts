export type Products = Product[]

export type Product = {
    id: number,
    name: string,
    description: string,
    price: number,
    image: string,
    category: string,
    stock: number,
    rating: number,
    brand?: string

}