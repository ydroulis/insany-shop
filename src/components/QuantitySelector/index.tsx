import React, { useRef, useState } from 'react';

import * as S from './styles';
import { Options } from '@/types/SelectorOptions';
import { FaChevronDown } from 'react-icons/fa';

interface QuantitySelectorProps {
    options: Options[]
    setValue: React.Dispatch<React.SetStateAction<string>>
    value: string
    label: string
    id: string
}

const QuantitySelector: React.FC<QuantitySelectorProps> = ({ options, setValue, value, label, id }) => {
    const [isOpen, setIsOpen] = useState(false);

    const handleClick = (option: string) => {
        setValue(option);
        setIsOpen(false);
    }

    return (
        <S.QuantityContainer>
            <label id={`${id}-label`} className="sr-only">
                {label}
            </label>
            <S.QuantitySelector
                id={id}
                role="combobox"
                aria-haspopup="listbox"
                aria-expanded={isOpen}
                aria-labelledby={`${id}-label`}
                aria-controls={`${id}-listbox`}
                onClick={() => setIsOpen(prev => !prev)}
            >
                <span>{value}</span>
                <FaChevronDown className='icon' size={16} color='#737380' aria-hidden="true" />
            </S.QuantitySelector>

            <S.QuantitySelectorMenu
                id={`${id}-listbox`}
                role="listbox"
                aria-labelledby={`${id}-label`}
                $isOpen={isOpen}
            >
                {options.map(option => (
                    <S.QuantitySelectorItem
                        key={option.value}
                        role="option"
                        tabIndex={0}
                        aria-selected={value === option.label}
                        onClick={() => handleClick(option.label)}
                        onKeyDown={(e) => {
                            if (e.key === 'Enter' || e.key === ' ') {
                                e.preventDefault();
                                handleClick(option.label);
                            }
                        }}
                    >
                        {option.label}
                    </S.QuantitySelectorItem>
                ))}
            </S.QuantitySelectorMenu>
        </S.QuantityContainer>
    );
}

export default QuantitySelector;