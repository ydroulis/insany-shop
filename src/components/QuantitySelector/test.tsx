import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import QuantitySelector from "./";
import { Options } from "@/types/SelectorOptions";

describe("<QuantitySelector />", () => {
    const mockOptions: Options[] = [
        { value: "1", label: "Option 1" },
        { value: "2", label: "Option 2" },
        { value: "3", label: "Option 3" },
    ];

    it("should render with initial value", () => {
        const setValue = jest.fn();
        render(
            <QuantitySelector
                options={mockOptions}
                value="Option 1"
                setValue={setValue}
                label="Selecione a quantidade"
                id="quantity"
            />
        );

        expect(
            screen.getByRole("combobox", { name: /selecione a quantidade/i })
        ).toBeInTheDocument();

        expect(screen.getAllByText("Option 1")).toHaveLength(2);

        expect(screen.getByRole("combobox").querySelector("svg")).toBeInTheDocument();
    });

    it("should open and close the listbox", () => {
        const setValue = jest.fn();
        render(
            <QuantitySelector
                options={mockOptions}
                value=""
                setValue={setValue}
                label="Selecione a quantidade"
                id="quantity"
            />
        );

        const button = screen.getByRole("combobox");

        expect(screen.queryByRole("listbox")).not.toBeInTheDocument();

        fireEvent.click(button);
        expect(screen.getByRole("listbox")).toBeVisible();

        fireEvent.click(button);
        expect(screen.queryByRole("listbox")).not.toBeInTheDocument();
    });

    it("should select an option with click", () => {
        const setValue = jest.fn();
        render(
            <QuantitySelector
                options={mockOptions}
                value=""
                setValue={setValue}
                label="Selecione a quantidade"
                id="quantity"
            />
        );

        const button = screen.getByRole("combobox");
        fireEvent.click(button);

        const option = screen.getByRole("option", { name: "Option 2" });
        fireEvent.click(option);

        expect(setValue).toHaveBeenCalledWith("Option 2");
    });

    it("should select option with Enter key", () => {
        const setValue = jest.fn();
        render(
            <QuantitySelector
                options={mockOptions}
                value=""
                setValue={setValue}
                label="Selecione a quantidade"
                id="quantity"
            />
        );

        const button = screen.getByRole("combobox");
        fireEvent.click(button);

        const option = screen.getByRole("option", { name: "Option 3" });
        fireEvent.keyDown(option, { key: "Enter", code: "Enter" });

        expect(setValue).toHaveBeenCalledWith("Option 3");
    });

    it("should select option with Space key", () => {
        const setValue = jest.fn();
        render(
            <QuantitySelector
                options={mockOptions}
                value=""
                setValue={setValue}
                label="Selecione a quantidade"
                id="quantity"
            />
        );

        const button = screen.getByRole("combobox");
        fireEvent.click(button);

        const option = screen.getByRole("option", { name: "Option 1" });
        fireEvent.keyDown(option, { key: " ", code: "Space" });

        expect(setValue).toHaveBeenCalledWith("Option 1");
    });

    it("should apply aria-selected correctly", () => {
        const setValue = jest.fn();
        render(
            <QuantitySelector
                options={mockOptions}
                value="Option 1"
                setValue={setValue}
                label="Selecione a quantidade"
                id="quantity"
            />
        );

        const button = screen.getByRole("combobox");
        fireEvent.click(button);

        const option1 = screen.getByRole("option", { name: "Option 1" });
        const option2 = screen.getByRole("option", { name: "Option 2" });

        expect(option1).toHaveAttribute("aria-selected", "true");
        expect(option2).toHaveAttribute("aria-selected", "false");
    });

    it("should match snapshot", () => {
        const { container } = render(
            <QuantitySelector
                options={mockOptions}
                value="Option 1"
                setValue={jest.fn()}
                label="Selecione a quantidade"
                id="quantity"
            />
        );
        expect(container).toMatchSnapshot();
    });
});
