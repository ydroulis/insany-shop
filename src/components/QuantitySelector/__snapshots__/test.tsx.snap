// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<QuantitySelector /> should match snapshot 1`] = `
<div>
  <div
    class="sc-bRKDuR fPjpkO"
  >
    <label
      class="sr-only"
      id="quantity-label"
    >
      Selecione a quantidade
    </label>
    <button
      aria-controls="quantity-listbox"
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="quantity-label"
      class="sc-hvigdm eqGjCL"
      id="quantity"
      role="combobox"
    >
      <span>
        Option 1
      </span>
      <svg
        aria-hidden="true"
        class="icon"
        color="#737380"
        fill="currentColor"
        height="16"
        stroke="currentColor"
        stroke-width="0"
        style="color: rgb(115, 115, 128);"
        viewBox="0 0 448 512"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        />
      </svg>
    </button>
    <div
      aria-labelledby="quantity-label"
      class="sc-fhHczv eXWIhS"
      id="quantity-listbox"
      role="listbox"
    >
      <button
        aria-selected="true"
        class="sc-ggWZvA gKPDnv"
        role="option"
        tabindex="0"
      >
        Option 1
      </button>
      <button
        aria-selected="false"
        class="sc-ggWZvA gKPDnv"
        role="option"
        tabindex="0"
      >
        Option 2
      </button>
      <button
        aria-selected="false"
        class="sc-ggWZvA gKPDnv"
        role="option"
        tabindex="0"
      >
        Option 3
      </button>
    </div>
  </div>
</div>
`;
